virtual_rainforest/core/registry.py

```diff
@@ -0,0 +1,222 @@
+    with resources.as_file(
+        resources.files(module) / schema_file_name
+    ) as schema_file_path:
+        register_schema(
+            module_name_short=module_name_short, schema_file_path=schema_file_path
+        )
+
+    # Find and register the constant dataclasses
+    register_constants_classes(module=module, module_name_short=module_name_short)
+
+
+def register_model(
+    module: ModuleType, module_name: str, module_name_short: str
+) -> None:
+    """Register the model component of a module.
+
+    This function automatically looks for a single subclass of a
+    :class:`~virtual_rainforest.core.base_model.BaseModel` within the members of the
+    provided module and adds it to the module registry.
+
+    Args:
+        module: A module object that might contain constants classes.
+        module_name: The full name of a module
+        module_name_short: The short name of the module.
+    """
+    from virtual_rainforest.core.base_model import BaseModel
+
+    # Extract BaseModel objects from the module root
+    model = [
+        (obj_name, obj)
+        for obj_name, obj in getmembers(module)
+        if isclass(obj) and issubclass(obj, BaseModel)
+    ]
+
+    # Trap missing and multiple models
+    if len(model) == 0:
+        excep = RuntimeError(f"Model object not found in {module_name}")
+        LOGGER.critical(excep)
+        raise excep
+
+    if len(model) > 1:
+        excep = RuntimeError(f"More than one model defined in in {module_name}")
+        LOGGER.critical(excep)
+        raise excep
+
+    # Trap models that do not follow the requirement that the BaseModel.model_name
+    # attribute matches the virtual_rainforest.models.model_name
+    # TODO - can we retire the model_name attribute if it just duplicates the module
+    #        name
+    model_name, model_obj = model[0]
+    if module_name_short != model_obj.model_name:
+        excep = RuntimeError(
+            f"The model_name attribute does not match the module name in {module_name}"
+        )
+        LOGGER.critical(excep)
+        raise excep
+
+    # Register the resulting single model class
+    MODULE_REGISTRY[module_name_short]["model"] = model_obj
+    LOGGER.info(f"Registered class for {module_name_short} model: {model_name}")
+
+
+def register_schema(module_name_short: str, schema_file_path: Path) -> None:
+    """Simple function to add configuration schema to the registry.
+
+    Args:
+        module_name: The name to register the schema under
+        schema_file_path: The file path to the JSON Schema file for the model
+
+    Raises:
+        ValueError: If the module name has already been used to register a schema
+    """
+
+    if (
+        module_name_short in MODULE_REGISTRY
+        and "schema" in MODULE_REGISTRY[module_name_short]
+    ):
+        excep = ValueError(
+            f"The module schema for {module_name_short} is already registered"
+        )
+        LOGGER.critical(excep)
+        raise excep
```

Comment:
A good point - this is a hangover from the separate registries, which I kept as a kind of guard against incomplete registration. But - that shouldn't happen unless someone tries to run the schema function independently of the register_submodule.

I think this all gets a lot cleaner if I can get the dataclass working, because then register_module can atomically add a module dataclass, and the deeper structure is guaranteed.